src/receivers/kafka/receiver.rs

```diff
@@ -358,52 +351,15 @@ impl KafkaReceiver {
     where
         T: serde::de::DeserializeOwned + prost::Message + Default,
     {
-        let request = match format {
+        match format {
             DeserializationFormat::Json => serde_json::from_slice::<T>(&data).map_err(|e| {
                 debug!("Failed to decode {}", e);
-                e
-            })?,
+                e.into()
+            }),
             DeserializationFormat::Protobuf => T::decode(Bytes::from(data)).map_err(|e| {
                 debug!("Failed to decode {}", e);
-                e
-            })?,
-        };
-        Ok(request)
-    }
-
-    // Helper method to send messages with cancellation support
-    async fn send_with_cancellation<T>(
```

Comment:
When you removed this in favor of the output.send() function it breaks clean shutdown of the system as output.send() can block indefinitely. If you compile and run and then send a SIGINT you'll notice that the receiver will not shutdown cleanly. It's important that we shutdown cleanly, flush any pending messages and commit offsets to Kafka before existing to ensure no data loss and reduce the frequency of duplicates. We need at-least-once semantics, however we also don't want to needlessly send duplicates if it can be avoided.